### ISSUE TITLE
Support PetSet mode

### ISSUE THREAD

**Issue by simonswine** [fabric8io/elasticsearch-cloud-kubernetes#44]
I am thinking about extending this plugin to support PetSets in its current 1.3 implementation. 

I think it would be great if we could support a safe way of scaling down. I would recon it's a good idea to have a hook during shutting down (after SIGTERM) like that:

- Look at my hostname which is (eg. es-data-N). If N+1 is smaller than desired PetSet size. Switch off data role and wait for all shards to be relocated. 

I am not really experienced with elasticsearch plugin development. But would a plugin solution like that be possible?

Another solution I could think of would be some go PID 1, that is doing that...

Any thoughts?
Someone else working on that?

**Comment by jimmidyson** [comment]
That's an interesting idea. I'm not too sure about petsets but if you resize a petset does it not terminate the spare petsets?

Your idea could work nicely with a PreStop hook & a longer graceful termination period perhaps?

**Comment by simonswine** [comment]
Yeah exactly I was thinking of a longer termination period. I am not familiar with ES and I guess it's hard to hook into the SIGTERM handling and migrate data away first. PreStop hook could work fine as well. I think I am going down that route for now, as the litte go helper could probably be used for other "Pets" as well. (And I am not really fond of setting up a Java environment :laughing: ). 

I guess I could rewrite the config and reload ES in such an event and then wait for data to be migrated of the local node...

I will let you know how that goes

**Comment by jimmidyson** [comment]
Cool! I haven't looked at the Elasticsearch API to see if there's a way to flush replicas or indeed if a simple shutdown would be enough - I assume it would as long as you have redundant replicas.

**Comment by simonswine** [comment]
Ok it seems to work pretty good with a lifecycle hook:
https://github.com/jetstack/elasticsearch-pet/blob/master/pre-stop-hook.sh

Maybe worth getting into the fabric8 image. I created a PR for kubernetes' charts:
https://github.com/kubernetes/charts/pull/58
